src/components/Modals/SmartHighlight.tsx

```diff
@@ -0,0 +1,128 @@
+
+  // close on click outside
+  useEffect(() => {
+    const clickHandler = ({ target }: MouseEvent) => {
+      if (!modal.current) return;
+      if (
+        !modalOpen ||
+        modal.current.contains(target) ||
+        trigger.current.contains(target)
+      )
+        return;
+      setModalOpen(false);
+    };
+    document.addEventListener("click", clickHandler);
+    return () => document.removeEventListener("click", clickHandler);
+  });
+
+  // close if the esc key is pressed
+  useEffect(() => {
+    const keyHandler = ({ keyCode }: KeyboardEvent) => {
+      if (!modalOpen || keyCode !== 27) return;
+      setModalOpen(false);
+    };
+    document.addEventListener("keydown", keyHandler);
+    return () => document.removeEventListener("keydown", keyHandler);
+  });
+
+  
+  const handleClick = () => {
+
+    let newMap = new Map<number, ColorType>();
+
+    rootWords.forEach((rootWord, index) => {
+        let descendantWordIds: number[] = [];
+        rootWord.descendants.forEach((word) => {
+            descendantWordIds.push(word.id)
+        });
+
+        let rootBlockColor: ColorType = {
+            colorFill: (index < RootColorPalette.length) ? RootColorPalette[index] : DEFAULT_COLOR_FILL,
+            textColor: (index < 10) ? '#000000' : (index < 20 || index >= RootColorPalette.length) ? DEFAULT_TEXT_COLOR : '#FFFFFF',
+            borderColor: "" // not used for root block
+        };
+
+        updateWordColor(ctxStudyId, descendantWordIds, ColorActionType.colorFill, rootBlockColor.colorFill);
+        updateWordColor(ctxStudyId, descendantWordIds, ColorActionType.textColor, rootBlockColor.textColor);
+        newMap.set(rootWord.descendants[0].strongNumber, rootBlockColor);
+    })
+    ctxSetRootsColorMap(newMap);
+    setModalOpen(false);
+    
+}
+
+  
+  return (
+    <>
+
+      <button
+        ref={trigger}
+        onClick={() => setModalOpen(!modalOpen)}
+        className="inline-flex items-center justify-center gap-2.5 rounded-full bg-primary px-8 py-4 text-center font-medium text-white hover:bg-opacity-90 lg:px-8 xl:px-10"
+      >
+        Smart Highlight
+      </button>
+
+      <div
+        className={`fixed left-0 top-0 z-999999 flex h-full min-h-screen w-full items-center justify-center bg-black/90 px-4 py-5 ${
+          modalOpen ? "block" : "hidden"
+        }`}
+      >
+        <div
+          ref={modal}
+          onFocus={() => setModalOpen(true)}
+          onBlur={() => setModalOpen(false)}
+          className="w-full max-w-142.5 rounded-lg bg-white px-8 py-12 text-center dark:bg-boxdark md:px-17.5 md:py-15"
+        >
+          <h3 className="pb-2 text-xl font-bold text-black dark:text-white sm:text-2xl">
+            Warning!
+          </h3>
+          <span className="mx-auto mb-6 inline-block h-1 w-22.5 rounded bg-primary"></span>
+          <p className="mb-10">
+            Using smart highlight will overwrite your custom styles on related words.
```

Comment:
It should be "identical words", not related words

Reply:
changed